src/main/java/com/hierynomus/msdfsc/ReferralCache.java

```diff
@@ -183,4 +170,46 @@ public String toString() {
 
     }
 
+
+    static class ReferralCacheNode {
+        String pathComponent;
+        Map<String, ReferralCacheNode> childNodes = new HashMap<>();
+        ReferralCacheEntry entry;
+
+        ReferralCacheNode(String pathComponent) {
+            this.pathComponent = pathComponent;
+        }
+
+        ReferralCacheNode(String pathComponent, ReferralCacheEntry entry) {
+            this.pathComponent = pathComponent;
+            this.entry = entry;
+        }
+
+        void addReferralEntry(Iterator<String> pathComponents, ReferralCacheEntry entry) {
+            if (pathComponents.hasNext()) {
+                String component = pathComponents.next();
```

Comment:
Should case normalisation (e.g., toLowerCase) be done here and in getReferralEntry to get case insensitive lookups? In general SMB is case preserving, so case sensitive caching might be a bit stricter than necessary. On the other hand it doesn't cause correctness issues, only potential redundant referral requests.

Reply:
What does the spec say? As far as I can see only the server is required to treat paths in a case-insensitive manner, nothing is said about the cache.

Reply:
Doesn't mention it afaik.